addons/godot-next/resources/resource_collections/resource_set.gd

```diff
@@ -15,72 +15,74 @@ class_name ResourceSet
 
 ##### PROPERTIES #####
 
-var _name: = ""
-var _type: Script = null
-
 var _data := {}
 
 ##### NOTIFICATIONS #####
 
+func _init() -> void:
+	resource_name = COLLECTION_NAME
+
+func _get(p_property: String):
+	if p_property.begins_with(PREFIX):
+		return _data.get(p_property.lstrip(PREFIX), null)
+	return null
+
+func _set(p_property: String, p_value) -> bool:
+	if p_property.begins_with(PREFIX):
+		var key = p_property.lstrip(PREFIX)
+		if not p_value:
+			#warning-ignore:return_value_discarded
+			_data.erase(key)
+			property_list_changed_notify()
+			return true
+		elif _data[key].get_script() == p_value.get_script():
+			var res = _instantiate_script(p_value) if p_value is Script else p_value
+			if res:
+				_data[key] = res
+		return true
+	return false
+
 func _get_property_list() -> Array:
-	return [ PropertyInfo.new_dictionary("_data", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_STORAGE).to_dict() ]
+	var list := []
+	if not _type:
+		return list
+	
+	list.append(PropertyInfo.new_group(PREFIX, PREFIX).to_dict())
+	list.append(PropertyInfo.new_dictionary(PREFIX, PROPERTY_HINT_RESOURCE_TYPE, "#%s" % _type.get_path(), PROPERTY_USAGE_DEFAULT).to_dict())
+	if _data.empty():
+		list.append(PropertyInfo.new_nil(PREFIX + EMPTY_ENTRY).to_dict())
+	for item in _data:
+		list.append(PropertyInfo.new_resource(PREFIX + item, "", PROPERTY_USAGE_EDITOR).to_dict())
+	
+	return list
 
 ##### OVERRIDES #####
 
-func can_contain(p_property: String) -> bool:
-	if not _name or not _type:
-		return false
-	return p_property.begins_with(_name + DELIMITER) and not p_property.ends_with(EMPTY_ENTRY)
-
-func add_element(p_script: Script) -> void:
+func _add_element(p_script: Script) -> void:
 	var key := ClassType.new(p_script).get_name()
 	if not _data.has(key):
 		_data[key] = p_script.new()
 
-func set_element(p_property: String, p_value) -> bool:
-	var key := p_property.lstrip(_name + DELIMITER)
-	if not p_value:
-		#warning-ignore:return_value_discarded
-		_data.erase(key)
-	else:
-		_data[key] = p_value
-	return true
-
-func get_element(p_property: String) -> Resource:
-	var key := p_property.lstrip(_name + DELIMITER)
-	return _data[key] if _data.has(key) else null
-
-func get_collection_property_list() -> Array:
-	var list = []
-	
-	if not _name or not _type:
-		return list
-	
-	var group_prefix = _name + DELIMITER
-	var item_prefix = _name + DELIMITER
-	list.append(PropertyInfo.new_group("%s %s" % [_name, COLLECTION_NAME], group_prefix).to_dict())
-	list.append(PropertyInfo.new_dictionary(_name, PROPERTY_HINT_RESOURCE_TYPE, "#%s" % _type.get_path(), PROPERTY_USAGE_EDITOR).to_dict())
-	
-	if _data.empty():
-		list.append(PropertyInfo.new_nil("%s%s" % [group_prefix, EMPTY_ENTRY]).to_dict())
-	for item in _data:
-		list.append(PropertyInfo.new_resource("%s%s" % [item_prefix, item], "", PROPERTY_USAGE_EDITOR).to_dict())
-	return list
+func _refresh_data() -> void:
+	if _type == null:
+		clear()
+		return
+	var data_keys := _data.keys()
+	for key in data_keys:
```

Comment:
`for key in data_keys:` -> `for a_key in _data.keys():`

Reply:
You never re-use data_keys, so there is no reason to cache it in a variable. Also, I recommend keeping temporary loop variables with a `a` or `an` prefix, to indicate their transitive nature.

Reply:
okay will keep that in mind :)